docs/features/patterns.md

```diff
@@ -0,0 +1,472 @@
+Pattern Matching for C#
+=======================
+
+Pattern matching extensions for C# enable many of the benefits of algebraic data types and pattern matching from functional languages, but in a way that smoothly integrates with the feel of the underlying language. The basic features are: [record types](records.md), which are types whose semantic meaning is described by the shape of the data; and pattern matching, which is a new expression form that enables extremely concise multilevel decomposition of these data types. Elements of this approach are inspired by related features in the programming languages [F#](http://www.msr-waypoint.net/pubs/79947/p29-syme.pdf "Extensible Pattern Matching Via a Lightweight Language") and [Scala](http://lampwww.epfl.ch/~emir/written/MatchingObjectsWithPatterns-TR.pdf "Matching Objects With Patterns").
+
+## Is Expression
+
+The `is` operator is extended to test an expression against a *pattern*.
+
+>*relational-expression*:
```

Comment:
It seems that [new lines](https://github.com/gafter/roslyn/blob/pattern-spec/docs/features/patterns.md) are not inserted between production rules correctly.